src/main/java/ru/yandex/clickhouse/BalancedClickhouseDataSource.java

```diff
@@ -0,0 +1,234 @@
+package ru.yandex.clickhouse;
+
+import org.slf4j.LoggerFactory;
+import ru.yandex.clickhouse.settings.ClickHouseProperties;
+
+import javax.sql.DataSource;
+import java.io.PrintWriter;
+import java.net.URISyntaxException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.*;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static ru.yandex.clickhouse.ClickhouseJdbcUrlParser.JDBC_CLICKHOUSE_PREFIX;
+
+public class BalancedClickhouseDataSource implements DataSource {
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(BalancedClickhouseDataSource.class);
+    private static final Pattern URL_TEMPLATE = Pattern.compile(JDBC_CLICKHOUSE_PREFIX + "//([a-zA-Z0-9_:,.]+)(/[a-zA-Z0-9_]+)?");
+
+    protected PrintWriter printWriter;
+    protected int loginTimeout = 0;
+    protected final ClickHouseDriver driver = new ClickHouseDriver();
+
+    private Random rnd = new Random();
+    private Set<String> disabledUrls = new HashSet<String>();
+    private Set<String> urls = new HashSet<String>();
```

Comment:
You have written:
`synchronized (urls) {
   ...
}`
but field "urls" is not final. It seems to me that is a mistake.